### PR TITLE
[quincy] RGW backports

### PR DESCRIPTION
**Pull request by soumyakoduri**
This PR contains backports of below fixes -

https://github.com/ceph/ceph/pull/50098
https://github.com/ceph/ceph/pull/50172
https://github.com/ceph/ceph/pull/51192
https://github.com/ceph/ceph/pull/50676
https://github.com/ceph/ceph/pull/51511
https://github.com/ceph/ceph/pull/49803
https://github.com/ceph/ceph/pull/52160

Fixes: https://tracker.ceph.com/issues/59674
Fixes: https://tracker.ceph.com/issues/59644
Fixes: https://tracker.ceph.com/issues/61391
Fixes: https://tracker.ceph.com/issues/61389
Fixes: https://tracker.ceph.com/issues/58897
Fixes: https://tracker.ceph.com/issues/61481
Fixes: https://tracker.ceph.com/issues/61877

Signed-off-by: Soumya Koduri <skoduri@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by soumyakoduri** [comment]
jenkins test submodules

**Comment by soumyakoduri** [comment]
jenkins test submodules

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:4717:81: error: 'key' is a protected member of 'rgw::sal::Object'
    ldpp_dout(dpp, 0) << __func__ << " ERROR: failed to transition obj(" << obj.key << ") read_mtime = " << read_mtime << " doesn't match mtime = " << mtime << dendl;
                                                                                ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_sal.h:861:17: note: declared protected here
    rgw_obj_key key;
                ^
```

**Comment by soumyakoduri** [comment]
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:4717:81: error: 'key' is a protected member of 'rgw::sal::Object'
>     ldpp_dout(dpp, 0) << __func__ << " ERROR: failed to transition obj(" << obj.key << ") read_mtime = " << read_mtime << " doesn't match mtime = " << mtime << dendl;
>                                                                                 ^
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_sal.h:861:17: note: declared protected here
>     rgw_obj_key key;
>                 ^
> ```

thanks. Fixed it.

**Comment by mkogan1** [comment]
@cbodley @soumyakoduri Think that this PR should also fix tracker: 
https://tracker.ceph.com/issues/65436 -- Getting Object Crashing radosgw services

**Comment by soumyakoduri** [comment]
> @cbodley @soumyakoduri Think that this PR should also fix tracker: https://tracker.ceph.com/issues/65436 -- Getting Object Crashing radosgw services

Sure @mkogan1 .. I will rebase and update the PR.

File: doc/radosgw/cloud-transition.rst [hunk 1]
```diff
@@ -302,6 +302,18 @@ Due to API limitations there is no way to preserve original object modification
    x-amz-meta-rgwx-source-mtime: 1608546349.757100363
    x-amz-meta-rgwx-versioned-epoch: 0
 
+In order to allow some cloud services detect the source and map the user-defined 'x-amz-meta-' attributes, below two additional new attributes are added to the objects being transitioned
```
Comment (anthonyeleven) [inline doc/radosgw/cloud-transition.rst:305]: s/some //

File: src/rgw/rgw_bucket_sync.cc [hunk 2]
```diff
@@ -557,6 +589,30 @@ void RGWBucketSyncFlowManager::init(const DoutPrefixProvider *dpp, const rgw_syn
   }
 }
 
+/*
+* These are the semantics to be followed while resolving the policy
```
Comment (anthonyeleven) [inline src/rgw/rgw_bucket_sync.cc:593]: Below are rules that are enforced when resolving policy conflicts:

File: src/rgw/rgw_bucket_sync.cc [hunk 3]
```diff
@@ -557,6 +589,30 @@ void RGWBucketSyncFlowManager::init(const DoutPrefixProvider *dpp, const rgw_syn
...
+*                         allowed         enabled
+*                         forbidden       disabled
+* allowed                 enabled         enabled
+*                         allowed         disabled
+*                         forbidden       disabled
+* forbidden               enabled         disabled
+*                         allowed         disabled
+*                         forbidden       disabled
+*
+* In case multiple group policies are set to reflect for any sync pair
```
Comment (anthonyeleven) [inline src/rgw/rgw_bucket_sync.cc:609]: If multiple group policies are set for any sync pair

**Review by anthonyeleven — COMMENTED** [review]
@zdover23 For the future

**Review by mkogan1 — APPROVED** [review]
LGTM, the SIGABRT does NOT occur after cherry-picking the commits in this PR on top of v17.2.7

**Pull request merged**